### DISCUSSION TITLE
Will the `"eslintConfig"` key in `package.json` files still work in ESLint v9?

### DISCUSSION THREAD

**Discussion by dustin-ruetz** [eslint/eslint#18131] [Q&A] [answered]
## Context

As of this writing using ESLint v8.56.0 I use the following configuration in my `package.json` files:

```json
{
  "scripts": {
    "lint": "eslint --config ./package.json",
    "lint:check": "npm run lint -- ./",
    "lint:fix": "npm run lint -- --fix ./"
  },
  "eslintConfig": {
    "extends": "./node_modules/dev-config/lib/eslint.config.js"
  }
}
```

Using the `eslintConfig.extends` key to point to the custom path of the `eslint.config.js` file is very helpful because it allows me to define the path to the configuration file once, and both the `eslint` CLI commands and the [ESLint VS Code extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) work as expected. As you can see by the path, this `eslint.config.js` file is published as part of a development config package that all of my other projects use.

## Documentation Contradictions

The reason I'm starting this discussion thread is because while the v8 [/docs/latest/use/configure/configuration-files](https://eslint.org/docs/latest/use/configure/configuration-files) page lists the `package.json` file as an option, the v9 [/docs/next/use/configure/configuration-files](https://eslint.org/docs/next/use/configure/configuration-files) page only mentions the `eslint.config.js` file and that it must be located in the root directory. With that being said, the v9 [/docs/next/use/command-line-interface#-c---config](https://eslint.org/docs/next/use/command-line-interface#-c---config) page makes it sound like `package.json` configuration _will_ still be supported:

> `-c`, `--config` example
>
> ```sh
> npx eslint -c ~/my-eslint.json file.js
> ```
>
> This example uses the configuration file at `~/my-eslint.json`.
>
> If `.eslintrc.*` and/or `package.json` files are also used for configuration (i.e., `--no-eslintrc` was not specified), the configurations are merged. Options from this configuration file have precedence over the options from `.eslintrc.*` and `package.json` files.

## Questions

So with all of this context presented, my question is:

1. Will the `"eslintConfig"` key in `package.json` files still work in ESLint v9?

And as a follow-on question:

2. If "yes" to the above, will the data structure change to match the flag config changes? I'm assuming that this would have to be a "yes" considering that v8's `extends` key will no longer be a config option as of v9, so perhaps the v9 implementation would look something like this?

```json
{
  "eslintConfig": ["./node_modules/dev-config/lib/eslint.config.js"]
}
```

**Comment by fasttime** [comment]
Configuration settings in package.json are not supported in the new config system (the default for v9). I guess the migration guide could be more explicit about that.

**Reply by dustin-ruetz** [reply]
Hi @fasttime, thanks for replying 🙏

Would you happen to have any insight on if the ESLint team would consider supporting the `"eslintConfig"` key in `package.json` files in v9?

To clarify:

1. I understand that it couldn't function the same way as it did in v8 due to A) the various API changes, and B) the fact that v9's flat config files now rely on a lot of JavaScript language features that can't be used in static JSON files (i.e. being able to `import` plugins, using the object spread syntax to extend `recommended` configuration settings, etc.).
1. As suggested in my original post, being able to set `"eslintConfig"` (either as a single string or an array of strings pointing to the path(s) of non-root configuration file(s)) would make intuitive sense; while we may not be able to write inline JSON configuration any more, it feels like we should be able to pass a path to custom location(s) of the flat config file(s).
1. I wouldn't consider a lack of support for the `"eslintConfig"` key as blocking the stable release of v9.

I can think of the following advantages for continuing to support a simplified (i.e. path only) implementation of this functionality:

1. Being able to specify a custom location of the configuration file is supported when running `eslint` as a CLI tool using the `--config` flag, and it feels like this is a valuable use case for the ESLint extension to support as well.
1. Supporting config files that are not located in the root of the repo can help keep repo roots cleaner, i.e. less config file clutter.
1. I've seen a lot of repos using the `"eslintConfig"` key in their `package.json` files.
1. A number of other popular tools (notably Jest, lint-staged and Prettier) continue to support `package.json` configuration.

**Reply by fasttime** [reply, marked as answer]
The main reason for removing the `"eslintConfig"` key lookup was to avoid the complexity of the legacy config system, where many ways existed to do the same thing. This is one of the aims stated in [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-1/) and is mentioned in [the original RFC](https://github.com/eslint/rfcs/blob/main/designs/2019-config-simplification/README.md#how-does-this-affect-configuration-via-packagejson).

Some of the initial decisions in the RFC have been revisited later because they turned out impractical or were posing challenges to the ecosystem, but it looks like this does not apply in your case. So I would not expect the team to consider reinstating support for the `"eslintConfig"` key in package.json files for the advantages you mention.

I think the best you could do is adding a eslint.config.js file to the root of your repo where you re-export the common config, i.e.

```js
// ./eslint.config.js
module.exports = require("dev-config/lib/eslint.config.js");
```

This should work also in vscode-eslint when they add support for ESLint v9.

**Reply by dustin-ruetz** [reply]
Thanks for providing this additional context @fasttime, I really appreciate it 🙏

I definitely understand the desire to simplify and streamline the implementation, and supporting a lot of configuration methods adds much more complexity.

---

All of my new codebases are now fully ESM, but I've tested the following with ESLint v9 and have confirmed that both the CLI tool and extension work with these changes:

```js
// eslint.config.js
export {default} from "dev-config/lib/eslint.config.js";
```

```json
// package.json
"scripts": {
  "lint:check": "eslint ./",
  "lint:fix": "eslint --fix ./"
},
```

---

If anyone is reading this discussion in the future, I did figure out another way to have the VS Code extension work with a non-root configuration file located in an arbitrary path. (This solution is specific to VS Code, but I'm sure other editors offer similar settings.) I'm not sure if this method will continue working when ESLint v9 reaches stable release status and/or if there will be further API changes being made to the extension, but I assume that @dbaeumer would know the most about this.

```json
// .vscode/settings.json
{
  "eslint.options": {
    "overrideConfigFile": "./node_modules/dev-config/lib/eslint.config.js"
  }
}
```

**Comment by mahnunchik** [comment]
@dustin-ruetz @fasttime 

It's very, very bad that the ESLint v9 does not support the `"eslintConfig"` key in the `package.json` file!

I use a very simple configuration in `package.json` with minor differences for directories:

```json
{
  "eslintConfig": {
    "extends": [
      "my-config-module"
    ],
    "env": {
      "browser": true
    },
    "overrides": [
      {
        "files": [
          "*.test.js",
          "test/*"
        ],
        "env": {
          "node": true,
          "mocha": true
        }
      }
    ]
  }
}
```

Why do I need to create **ANOTHER config file** `eslint.config.*` in the root directory? Everything worked fine with the `"eslintConfig"` key in the `package.json` file.

Yes, I am not against renaming the key to `"eslintFlatConfig"`, just leave support for the configuration without creating **more and more files** in the root directory.

**Reply by fasttime** [reply]
> Why do I need to create **ANOTHER config file** `eslint.config.*` in the root directory? Everything worked fine with the `"eslintConfig"` key in the `package.json` file.

The main reason is that it's not generally possible in ESLint v9, like it was possible in ESLint v8, to write a config file in JSON format.

For example, there is no longer a standard to resolve a shared config by name by specifying `"my-config-module"` in the `extends` field. Instead, in ESLint v9 you have to import plugins by yourself and include them in the config array. This gives you more control because you can customize your whole config programmatically, but it also means that static formats like JSON don't work any more.

We had some discussions about adding support for JSON configs with a limited set of features specifically for monorepos in https://github.com/eslint/eslint/issues/18385#issuecomment-2104749434, but I think this is not your use case.

**Reply by mahnunchik** [reply]
Everything is very strange with this flat configuration format https://github.com/eslint/eslint/issues/19041#issuecomment-2432830047

**Comment by reverofevil** [comment]
This sounds like a problem. There's clearly a need to set up eslint config for a monorepo module without cluttering code of the module with a config that will only have a single reexport.

Flat config system was partly implemented in order to undo the mess with configs-as-nodemodules, so maintainers definitely understand they're needed, but somehow it's still impossible to do just something in the lines of `eslint -c @foo/bar`.
